### PR TITLE
Migrating code from Client Library and Engine

### PR DESCRIPTION
**Pull request by Sajjon**
Migrating code from Client Library and Engine.

### CODE REVIEW

File: build.gradle
Comment (flotothemoon) [inline build.gradle:90]: Why this change?

File: src/main/java/com/radixdlt/crypto/ECKeyUtils.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECKeyUtils.java:82]: Why is it relevant what version of BouncyCastle Android has by default?

File: src/main/java/com/radixdlt/crypto/Hash.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/Hash.java:180]: Not a fan of this change since getID is the (unfortunate) convention across a lot of our codebase, so this would be inconsistent.

File: src/main/java/com/radixdlt/crypto/ECDSASignature.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECDSASignature.java]: Should add an explanatory message to this exception

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECKeyPair.java:118]: unnecessary newline

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECKeyPair.java]: I don't think these comments add any value in the source. If anything, you can make them here to explain to reviewers, but future maintainers don't need to know this

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECKeyPair.java]: Exception thrown here but caught a few lines later in the catch-all.

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECKeyPair.java]: Should replace with either proper log or rethrow if severe

File: src/main/java/com/radixdlt/crypto/ECKeyUtils.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECKeyUtils.java:126]: Is `lastByte == 1` a valid value then?

File: src/main/java/com/radixdlt/crypto/ECMultiplicationScalar.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECMultiplicationScalar.java]: Confused copyright notice here (note the double asterisks)

File: src/main/java/com/radixdlt/crypto/Hash.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/Hash.java]: extra "the" in "use *the* any of the"

File: src/main/java/com/radixdlt/crypto/Hash.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/Hash.java:124]: Not strictly required, but we usually explicitly check for null first using 
```
Objects.requireNonNull(alreadyHashedData, "alreadyHashedData");
```

File: src/main/java/com/radixdlt/crypto/HashHandler.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/HashHandler.java:4]: Incorrect license format, similar to above except here it's a missing whitespace (what?)

File: src/main/java/com/radixdlt/crypto/KeyHandler.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/KeyHandler.java]: `beDeterministic` sounds awkward to me, could just be `deterministic`

File: src/main/java/com/radixdlt/crypto/KeyHandler.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/KeyHandler.java]: Would be cleaner if these were just inlines, imo this actually makes it harder to read (though I appreciate what you're trying to do here)

File: src/main/java/com/radixdlt/crypto/SHAHashHandler.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/SHAHashHandler.java]: Should throw IllegalStateException instead of just a comment

File: src/main/java/com/radixdlt/crypto/Signatures.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/Signatures.java]: could be shortened to "minimumValidSignatures"

File: src/main/java/com/radixdlt/crypto/encryption/Crypt.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/encryption/Crypt.java]: funky license header again

File: src/main/java/com/radixdlt/crypto/encryption/CryptOperation.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/encryption/CryptOperation.java]: Same here.

File: src/main/java/com/radixdlt/crypto/encryption/CryptOperation.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/encryption/CryptOperation.java:20]: Why do we need the boolean field here?
If really necessary, isEncryption could be implemented as just 
```

public boolean isEncryption() {
	return this == ENCRYPT;
}
```

File: src/main/java/com/radixdlt/crypto/encryption/ECIES.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/encryption/ECIES.java]: License header.

File: src/main/java/com/radixdlt/crypto/encryption/ECIES.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/encryption/ECIES.java]: Do we really need to re-create an expensive SecureRandom instance every time this is called?

File: src/main/java/com/radixdlt/crypto/encryption/EncryptedPrivateKey.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/encryption/EncryptedPrivateKey.java]: License header. Please check all the others, I don't want to spam all of them.

File: src/main/java/com/radixdlt/crypto/encryption/Encryptor.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/encryption/Encryptor.java:18]: .. missing license header

File: src/main/java/com/radixdlt/crypto/encryption/PrivateKeyEncrypter.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/encryption/PrivateKeyEncrypter.java]: Could use null check

Comment (flotothemoon) [reply src/main/java/com/radixdlt/crypto/encryption/PrivateKeyEncrypter.java]: Actually, all methods here would benefit from a null check

File: src/main/java/com/radixdlt/crypto/keystore/Crypto.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/keystore/Crypto.java:29]: Spacing between individual members would make this easier to read

File: src/main/java/com/radixdlt/crypto/keystore/Keystore.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/keystore/Keystore.java:29]: Add space

File: src/main/java/com/radixdlt/crypto/keystore/Pbkdfparams.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/keystore/Pbkdfparams.java:29]: Same here

File: src/main/java/com/radixdlt/identifiers/RadixAddress.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/identifiers/RadixAddress.java]: Same comment about this comment as above, don't think it's relevant.

File: src/main/java/com/radixdlt/identifiers/RadixAddress.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/identifiers/RadixAddress.java:135]: Same here.

File: src/test/java/com/radixdlt/crypto/ECKeyPairTest.java
Comment (flotothemoon) [inline src/test/java/com/radixdlt/crypto/ECKeyPairTest.java]: Comment as above.

File: src/test/java/com/radixdlt/crypto/HashTest.java
Comment (flotothemoon) [inline src/test/java/com/radixdlt/crypto/HashTest.java:118]: Weird spacing going on in this file

**Review by flotothemoon — CHANGES_REQUESTED** [review]
Good first go. There are some questions regarding the strategy of these PRS as well as issues with the handling of failure cases, formatting and idiomatic use across these PRs that should be addressed before merging.

File: src/main/java/com/radixdlt/crypto/ECKeyUtils.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECKeyUtils.java:82]: This is copied over from Java Library, which considered Android. 

Legend:
Let `Y <- X` denote "X is dependent on Y"
`JC`: Java Common
`JL`: Java Library
`AW`: Android Wallet

We have (before the notion of Java Common):
`JL <- AW`

And I'm introducing:
`JC  <- JL <- AW`,

So what ever code we had in `JL` to support `AW`, is propagated into `JC`, does that make sense?

Or should that code, in fact, never ever have been part of `JL` to begin with? I.e. should just be removed?

File: src/main/java/com/radixdlt/crypto/Hash.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/Hash.java:180]: OK, hmm, this is contrary to instructions I got from @clstrfsck , which told be to drop "get" prefix for getter, if you remember Martin? Or did that only apply to "pure instance variable getters" and not derived values (being dependent on instance variables, but themselves calculated (and potentially cached))?

File: src/main/java/com/radixdlt/crypto/ECDSASignature.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECDSASignature.java]: good point!

File: src/main/java/com/radixdlt/identifiers/RadixAddress.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/identifiers/RadixAddress.java]: Any comments with "MIGRATION" is to be removed by me :D I was gonna do it "last minute", just there to help me keep track of things, but yes, you are right, I should not forget to remove them :)

File: src/main/java/com/radixdlt/identifiers/RadixAddress.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/identifiers/RadixAddress.java:135]: to be removed

File: src/test/java/com/radixdlt/crypto/ECKeyPairTest.java
Comment (Sajjon) [reply src/test/java/com/radixdlt/crypto/ECKeyPairTest.java]: to be removed

File: src/test/java/com/radixdlt/crypto/HashTest.java
Comment (Sajjon) [reply src/test/java/com/radixdlt/crypto/HashTest.java:118]: yeah IntelliJ is freaking out sometimes... hmm...  will investigate.

File: build.gradle
Comment (Sajjon) [reply build.gradle:90]: `api` is transient, where `compile` is not (also, [`compile` is deprecated, so if we don't wanna use the transient `api`, we should migrate to `implementation`](https://stackoverflow.com/a/44419574/1311272)), If I understand it correctly, this will make sure that we are using the same dependency (in this case, the line you commented on: `guava`) across Java Common, Engine Library, Java Library and any code using Java Library (e.g. Android Wallet).

File: src/main/java/com/radixdlt/crypto/keystore/Crypto.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/keystore/Crypto.java:29]: Agreed! This was just copied over, adding newlines after each instance variable 👍

File: src/main/java/com/radixdlt/crypto/encryption/PrivateKeyEncrypter.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/encryption/PrivateKeyEncrypter.java]: @flotothemoon Hmm like this:

```java
    public static String createEncryptedPrivateKey(String password) throws GeneralSecurityException {
        ECKeyPair ecKeyPair = new ECKeyPair();
        String privateKey = Objects.requireNonNull(Bytes.toHexString(ecKeyPair.getPrivateKey()));
        byte[] salt = getSalt().getBytes(StandardCharsets.UTF_8);

        SecretKey derivedKey = Objects.requireNonNull(getSecretKey(password, salt, ITERATIONS, KEY_LENGTH));

        Cipher cipher = Objects.requireNonNull(Cipher.getInstance("AES/CTR/NoPadding"));
        cipher.init(Cipher.ENCRYPT_MODE, derivedKey);
        byte[] iv = cipher.getParameters().getParameterSpec(IvParameterSpec.class).getIV();

        String cipherText = encrypt(cipher, privateKey);
        byte[] mac = generateMac(derivedKey.getEncoded(), Bytes.fromHexString(cipherText));

        Keystore keystore = Objects.requireNonNull(createKeystore(ecKeyPair, cipherText, mac, iv, salt));

        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        return Objects.requireNonNull(gson.toJson(keystore));
    }
```

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECKeyPair.java]: to be removed

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECKeyPair.java]: yeah most code here is just copy paste over, some of your comments will not be addressed, especially code related to crypto, because in coming PR I will make some changes and improvements to the crypto related code. See [RPNV1-62](https://radixdlt.atlassian.net/browse/RPNV1-62).

File: src/main/java/com/radixdlt/crypto/ECKeyUtils.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECKeyUtils.java:126]: yes, that is the case of PrivateKey of `1`, which is a perfectly valid PrivateKey and used by many test vectors, hence the need for change (and improvement). I use some common secp256k1 test vectors for deterministic signatures using the private key of `1` :).

Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECKeyUtils.java:126]: All scalars on interval `1..<N`, where `N` is the order of any elliptic curve is a perfectly valid private key.

File: src/main/java/com/radixdlt/crypto/ECMultiplicationScalar.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECMultiplicationScalar.java]: well spotted! I must have messed something up, thx!

File: src/main/java/com/radixdlt/crypto/Hash.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/Hash.java]: nice spotted!

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECKeyPair.java]: I changed my mind! Will address these things now, but might be change later anyway. Thanks for pointing this out, improving this.

File: src/main/java/com/radixdlt/crypto/HashHandler.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/HashHandler.java:4]: yeah strange, I must have messed something up, found in 16 places, found and replaced :D

File: src/main/java/com/radixdlt/crypto/KeyHandler.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/KeyHandler.java]: Changed to `useDeterministicSignatures` :D also added java doc for the two boolean params.

File: src/main/java/com/radixdlt/crypto/SHAHashHandler.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/SHAHashHandler.java]: Ah my bad, this constructor is in fact used (so removing comment.)

File: src/main/java/com/radixdlt/crypto/Signatures.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/Signatures.java]: changed.

File: src/main/java/com/radixdlt/crypto/encryption/CryptOperation.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/encryption/CryptOperation.java:20]: Yup! Thx!

File: src/main/java/com/radixdlt/crypto/KeyHandler.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/KeyHandler.java]: yeah using an IDE the usage of local variables just clutters the code. But not using an IDE I wanted to highlight what the boolean values meant, because Java is a crappy language without argument labels :D, but sure, I will remove them.

Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/KeyHandler.java]: Adding documentation as well.

File: src/main/java/com/radixdlt/crypto/encryption/ECIES.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/encryption/ECIES.java]: good catch, fixed, also added throwing of `IllegalState` in the constructor.

File: src/main/java/com/radixdlt/crypto/encryption/Encryptor.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/encryption/Encryptor.java:18]: added

File: src/main/java/com/radixdlt/crypto/encryption/PrivateKeyEncrypter.java
Comment (flotothemoon) [reply src/main/java/com/radixdlt/crypto/encryption/PrivateKeyEncrypter.java]: No that's not what I meant. 
I meant checking the input arguments to the method (password)

Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/encryption/PrivateKeyEncrypter.java]: ah, roger that!

File: build.gradle
Comment (clstrfsck) [reply build.gradle:90]: ^^ What he said, but transient -> transitive

File: src/main/java/com/radixdlt/crypto/Hash.java
Comment (clstrfsck) [reply src/main/java/com/radixdlt/crypto/Hash.java:180]: I think there is a balance here between aesthetics and the effort to review all the places the code changes.  I think we have this sorted out now in any case?

File: src/main/java/com/radixdlt/crypto/ECKeyUtils.java
Comment (clstrfsck) [reply src/main/java/com/radixdlt/crypto/ECKeyUtils.java:82]: Hmm.  Not an Android expert, but suggest we leave this in place until we can find out whether it is a good idea to remove.  From my sketchy knowledge of crypto primitives on Android, the BouncyCastle provider on Android has been resolved to some degree in recent releases, but it may be that we can support more versions of Android if we retain this.

File: build.gradle
Comment (Sajjon) [reply build.gradle:90]: Ops, hehe, right! Transient is when a property is not saved in database!

File: src/main/java/com/radixdlt/crypto/ECDSASignature.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECDSASignature.java]: Since only the happy path returns anything, these variables and the return statement can be compressed into the `try` block

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECKeyPair.java]: Should add explanatory message

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/ECKeyPair.java]: Should use try-with-resources here to ensure stream is closed (which it isn't right now)

File: src/main/java/com/radixdlt/crypto/encryption/Encryptor.java
Comment (flotothemoon) [inline src/main/java/com/radixdlt/crypto/encryption/Encryptor.java]: What's this about? Is this still relevant?

File: src/main/java/com/radixdlt/crypto/ECDSASignature.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECDSASignature.java]: nice! good improvement!

File: src/main/java/com/radixdlt/crypto/encryption/Encryptor.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/encryption/Encryptor.java]: not sure, not my code, copied over.

I can remove the `TODO`, I understand what the code does of course, but cannot come up with a better solution (I think this is Josh's code, and apparently neither could he (at that time
))

Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/encryption/Encryptor.java]: added comment clairifiy the code, like so:

```java
public byte[] decrypt(byte[] data, ECKeyPair accessor) throws CryptoException {
  for (EncryptedPrivateKey protector : protectors) {
    try {
      return accessor.decrypt(data, protector);
    } catch (CryptoException e) {
      // exception could so that we can proceed with next `protector`
    }
  }

  throw new CryptoException("Unable to decrypt any of the " + protectors.size() + " protectors.");
}
```

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECKeyPair.java]: Thx! @clstrfsck provided me with a simplified improved implementation that I copied over.

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java
Comment (Sajjon) [reply src/main/java/com/radixdlt/crypto/ECKeyPair.java]: yupp! fixed!.

**Review by flotothemoon — APPROVED** [review]

**Review by clstrfsck — APPROVED** [review]
LGTM

**Pull request merged**
